endpoint: Update proxy policies when applying policy map changes out-of-band

jrajahalme opened the pull request:
Update proxy policies in addition to bpf maps when applying policy map changes outside of the full endpoint regeneration logic.  Full regeneration is done when policy changes, without which there are no added or deleted proxy redirects. Given this we only need to update the proxy policy on this code path.  Proxy updates are only done if the changed bpf map entries have a non-zero proxy port.  The callers are expected to provide a context which bounds the wait time on proxy
completions if any.

Testing the problem with #9208 turned out to be very challenging. Basically it seems the proxy gets the policy update due to the full regeneration before it gets the request, but as full regeneration is currently gated at at most 2 times per second, this behavior cannot be relied upon, so this fix should still be applied.

Fixes: #9208

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9243)
<!-- Reviewable:end -->

raybejjani commented on pkg/endpointmanager/manager.go:
```diff
@@ -74,24 +75,51 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
+
 	eps := mgr.GetEndpoints()
-	wg.Add(len(eps))
+	wg.Add(len(eps) + 1)
```
It might be worth explaining that the +1 is for the proxy further down. Not a big deal, though

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -825,7 +825,10 @@ func (e *Endpoint) SkipStateClean() {
+// The bool pointed by hadProxy, if not nil, will be set to 'true' if
+// the deleted entry had a proxy port assigned to it.  *hadProxy is
+// not otherwise changed (e.g., it is never set to 'false').
+func (e *Endpoint) deletePolicyKey(keyToDelete policy.Key, incremental bool, hadProxy *bool) bool {
```
I'm surprised that `hadProxy` is passed in as a pointer instead of returning it. I'm guessing changing the return means changing the callsites for `deletePolicyKey` in a bigger way? (since the ->true accumulation would have to happen at outside here).

jrajahalme replied:
Right, it is not strictly a return parameter, and this way the PR become simpler.

ianvernon commented on pkg/endpointmanager/manager.go:
```diff
@@ -74,24 +75,51 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
 }
 
+// waitForProxyCompletions blocks until all proxy changes have been completed.
+func (mgr *EndpointManager) waitForProxyCompletions(proxyWaitGroup *completion.WaitGroup) error {
```
This function does not need the `*EndpointManager` receiver.

jrajahalme replied:
Removed :-)

ianvernon commented on pkg/endpointmanager/manager.go:
```diff
@@ -74,24 +75,51 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
 
+	go func() {
+		if err := mgr.waitForProxyCompletions(proxyWaitGroup); err != nil {
+			log.Warning("Failed to apply L7 proxy policy changes. These will be re-applied in future updates.")
```
Use `log.WithError(err)...`

jrajahalme replied:
Done.

jrajahalme committed:
test: Add L3-dependent L7 test with toFQDN

Add an L3-dependent L7 test with toFQDN.

Clear the FQDN cache between tests.

Properly prevent curl from resolving DNS names.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Update proxy policies when applying policy map changes out-of-band

Update proxy policies in addition to bpf maps when applying policy map
changes outside of the full endpoint regeneration logic.  Full
regeneration is done when policy changes, without which there are no
added or deleted proxy redirects. Given this we only need to update
the proxy policy on this code path.  Proxy updates are only done if
the changed bpf map entries have a non-zero proxy port.  The callers
are expected to provide a context which bounds the wait time on proxy
completions if any.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
